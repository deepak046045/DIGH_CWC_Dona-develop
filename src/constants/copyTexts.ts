import mockUserData from '@/mocks/mockUserData.json';
import { AppointmentTypes } from '@/shared/enums/AppointmentTypes';

export const copyTexts = {
  WelcomeSection: {
    welcomeHeaderTextUngated: `Welcome to Kisunla support.`,
    welcomeHeaderTextGated: `Hi ${mockUserData.firstName}, welcome to Kisunla support`,
    welcomeSubHeaderTextUngated: `Support resources for your treatment journey. Register to set up a
      calendar of appointments, set up reminders, and assign care
      partners.`,
    welcomeSubHeaderTextGated: `Support resources for your treatment journey. Set up your calendar of appointments, reminders, and care partners.`,
  },
  MyAppointments: {
    header: 'My appointments',
    subHeader:
      'Appointment reminders will be sent to you and your care partners 5 days and 1 day before your scheduled appointment.',
    noAppointments: {
      firstPiece: 'No appointments saved.',
      link: 'Add New Appointments',
      secondPiece: 'to get started.',
    },
    noUpcomingAppointments: {
      firstPiece: 'You have no upcoming appointments.',
      addLink: 'Add new appointments',
      or: 'or',
      viewLink: 'view all Appointments',
    },
    tableColumnHeaders: {
      date: 'Date',
      time: 'Time',
      address: 'Address',
      appointmentType: 'Type of appointment',
      carePartners: 'Care partner(s)',
      actions: 'Actions',
    },
    buttons: {
      newAppointment: 'New appointment',
      edit: 'Edit',
      delete: 'Delete',
    },
    deleteModal: {
      header: 'Are you sure you want to delete this appointment?',
      subHeader: 'This cannot be undone.',
      primaryButtonLabel: 'Yes, delete',
      secondaryButtonLabel: 'Cancel',
    },
    deleteAlert: {
      message: 'The appointment has been successfully deleted.',
    },
    appointmentTypes: {
      [AppointmentTypes.INFUSION]: 'IV infusion Dose',
      [AppointmentTypes.GENERAL_DOCTOR]: 'General Practicioner Check-in',
      [AppointmentTypes.SAFETY_MRI]: 'Safety MRI',
      [AppointmentTypes.NEUROLOGIST]: 'Neurologist Check-in',
    },
  },
  TreatmentIntro: {
    titleGated: 'My treatment plan',
    titleUngated: 'Recommended treatment',
    subtitleUngated:
      'Setting you up for success in your treatment. Setting you up for success in your treatment.',
    textUngated:
      'This treatment map serves as a suggested guide. Always follow your doctor’s directions. Insurance approval may be needed or the start of your infusion therapy may be delayed.',
    addApointmentBtnGated: 'Add Appointment',
  },
  TreatmentDownload: {
    reminderUngated:
      'Always follow your doctor’s directions. Insurance approval may be needed or the start of your infusion therapy may be delayed.',
    downloadUngated:
      'For a printable version of the treatment map download the following.',
    downloadGated: 'Print or save your Treatment Plan as a PDF',
    downloadButtonUngated: 'Treatment Map PDF',
    downloadButtonGated: 'Print My Plan',
  },
  Header: {
    signIn: 'Sign in',
    signUp: 'Sign up',
    signOut: 'Sign out',
    viewAppointments: 'View your appointments',
    nextAppointmentToday: 'Your next appointment is today.',
    nextAppointmentIsIn: 'Your next appointment is in',
    day: 'day',
    days: 'days',
  },
  Footer: {
    primaryLinks: {
      prescribingInformation: 'Prescribing Information',
      medicationGuide: 'Medication Guide',
    },
    globalLinks: {
      privacyPolicy: 'Privacy policy',
      termsOfUse: 'Terms of use',
      accessibilityStatement: 'Accessibility Statement',
    },
    saveToDeviceHomeButton: 'Save to device home',
  },
  AddAppointment: {
    title: 'Add Appointment',
    subtitle:
      'Appointment reminders will be sent to you and your care partners 5 days and 1 day before your scheduled appointment.',
    selectOptionLabels: {
      [AppointmentTypes.INFUSION]: 'Infusion',
      [AppointmentTypes.NEUROLOGIST]: 'Neurologist',
      [AppointmentTypes.SAFETY_MRI]: 'Safety MRI',
      [AppointmentTypes.GENERAL_DOCTOR]: 'General Doctor',
    },
    labels: {
      appointmentType: 'TYPE OF APPOINTMENT',
      infusionNumber: 'INFUSION # (OPTIONAL)',
      address: 'ADDRESS',
      date: 'DATE',
      time: 'TIME',
      carePartners: 'CARE PARTNERS (OPTIONAL)',
    },
    hints: {
      date: 'Format: MM/DD/YYYY',
      time: 'Format: HH:MM',
    },
    formErrors: {
      appointmentType: 'Select an appointment type',
      fieldRequired: 'This field is required',
      addressExceedLimit: 'Address exceeds 150 characters', // TODO: define this copy
      invalidDate: 'Date is invalid',
      invalidTime: 'Time is invalid',
    },
    confirmationModal: {
      title: 'Your appointment has been saved.',
      primaryButtonLabel: 'View my appointments',
      secondaryButtonLabel: 'Add new appointment',
    },
    addCarePartnerBtn: 'Add new care partner',
    saveBtn: 'Save',
    cancelBtn: 'Cancel',
  },
  AddCarePartnerModal: {
    title: 'Add care partner',
    name: 'NAME',
    email: 'EMAIL ADDRESS',
    cancelButton: 'Cancel',
    saveButton: 'Save',
    notSavedWarning: 'You have not saved your care partner.',
    allDetailsLost: 'All details will be lost.',
    closeAnywayButton: 'Close anyway',
    formValidation: {
      requiredField: 'This field is required',
      invalidEmail: 'Invalid email address',
      nameLength: 'Name must not be longer than 50 characters',
      emailLength: 'Email address must not be longer than 80 characters',
    },
  },
};
