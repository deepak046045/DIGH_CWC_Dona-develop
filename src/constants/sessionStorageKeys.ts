export const sessionStorageKeys = {
  IsUpcomingAppointmentAlertDismissed: 'isUpcomingAppointmentBannerDismissed',
};
