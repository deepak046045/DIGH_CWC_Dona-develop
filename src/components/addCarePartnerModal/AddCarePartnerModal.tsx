import React, { useEffect, useRef, useState } from 'react';
import {
  LdsModal,
  LdsButton,
  LdsTextField,
  LdsValidationError,
  LdsIcon,
  LdsAlert,
} from '@elilillyco/ux-lds-react';
import modal from '@/scss/themes/modal.module.scss';
import buttons from '@/scss/themes/buttons.module.scss';
import { copyTexts } from '@/constants/copyTexts';
import { Formik, FormikProps, FormikHelpers } from 'formik';
import {
  AddCarePartnerForm,
  AddCarePartnerInitValues,
} from '@/forms/addCarePartnerForm/AddCarePartnerForm';
import { AddCarePartnerValidationSchema } from '@/forms/addCarePartnerForm/AddCarePartnerFormValidation';
import styles from './AddCarePartnerModal.module.scss';

type AddCarePartnerModalProps = {
  isModalOpen: boolean;
  setIsModalOpen: (arg: boolean) => void;
  onSubmit: (values: AddCarePartnerForm) => void;
};

export function AddCarePartnerModal({
  isModalOpen,
  setIsModalOpen,
  onSubmit,
}: AddCarePartnerModalProps) {
  const formikRef = useRef<FormikProps<AddCarePartnerForm>>(null);
  const [isConfirmClose, setIsConfirmClose] = useState(false);

  const handleCloseModal = () => {
    if (formikRef.current?.dirty && !isConfirmClose) {
      setIsConfirmClose(true);
    } else {
      setIsModalOpen(false);
      setIsConfirmClose(false);
    }
  };

  const formSubmit = (
    values: AddCarePartnerForm,
    { setSubmitting }: FormikHelpers<AddCarePartnerForm>
  ) => {
    onSubmit(values);
    setIsConfirmClose(false);
    setSubmitting(false);
  };

  useEffect(() => {
    if (isModalOpen && formikRef.current) {
      formikRef.current.resetForm();
    }
  }, [isModalOpen]);

  return (
    <div className={`${modal.primary} ${styles.modal}`}>
      <LdsModal
        modalId="addCarePartnerModal"
        open={isModalOpen}
        setModalOpen={setIsModalOpen}
        persistent
      >
        {isConfirmClose ? (
          <div
            className={styles.notSavedAlert}
            data-testid="care-partner-close-warning"
          >
            <LdsAlert level="warning" inline>
              <div className={styles.notSavedAlertContent}>
                <div>
                  <strong>
                    {copyTexts.AddCarePartnerModal.notSavedWarning}
                  </strong>
                  &nbsp;{copyTexts.AddCarePartnerModal.allDetailsLost}
                </div>
                <LdsButton
                  classes={`${buttons.secondary} ${styles.closeAnywayButton}`}
                  onClick={handleCloseModal}
                  data-testid="care-partner-close-anyway-btn"
                >
                  {copyTexts.AddCarePartnerModal.closeAnywayButton}
                </LdsButton>
              </div>
            </LdsAlert>
          </div>
        ) : (
          <div className={styles.closeModalButtonWrapper}>
            <LdsButton
              onClick={handleCloseModal}
              clearDefaultClasses
              className={styles.closeModalButton}
            >
              <LdsIcon name="X" className={styles.closeIcon} />
            </LdsButton>
          </div>
        )}
        <h2 className={modal.modalTitle}>
          {copyTexts.AddCarePartnerModal.title}
        </h2>
        <Formik
          innerRef={formikRef}
          initialValues={AddCarePartnerInitValues}
          validationSchema={AddCarePartnerValidationSchema}
          onSubmit={formSubmit}
        >
          {({ handleChange, handleSubmit, values, errors, touched }) => (
            <form onSubmit={handleSubmit} className={styles.form}>
              <div className={styles.formContent}>
                <div>
                  <LdsTextField
                    id="name"
                    label={copyTexts.AddCarePartnerModal.name}
                    name="name"
                    value={values.name}
                    onChange={handleChange}
                    error={errors.name && touched.name}
                  />
                  {errors.name && touched.name && (
                    <LdsValidationError id="errName">
                      {errors.name}
                    </LdsValidationError>
                  )}
                </div>
                <div>
                  <LdsTextField
                    id="emailAddress"
                    label={copyTexts.AddCarePartnerModal.email}
                    name="emailAddress"
                    value={values.emailAddress}
                    onChange={handleChange}
                    error={errors.emailAddress && touched.emailAddress}
                  />
                  {errors.emailAddress && touched.emailAddress && (
                    <LdsValidationError id="errEmailAddress">
                      {errors.emailAddress}
                    </LdsValidationError>
                  )}
                </div>
              </div>
              <div className={modal.modalActions}>
                <LdsButton
                  classes={`${buttons.secondary}`}
                  onClick={handleCloseModal}
                >
                  {copyTexts.AddCarePartnerModal.cancelButton}
                </LdsButton>
                <LdsButton
                  type="submit"
                  classes={`${buttons.primary}`}
                  data-testid="add-care-partner-save-button"
                >
                  {copyTexts.AddCarePartnerModal.saveButton}
                </LdsButton>
              </div>
            </form>
          )}
        </Formik>
      </LdsModal>
    </div>
  );
}
