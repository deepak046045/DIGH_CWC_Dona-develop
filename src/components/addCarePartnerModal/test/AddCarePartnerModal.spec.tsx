import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { copyTexts } from '@/constants/copyTexts';
import { AddCarePartnerModal } from '../AddCarePartnerModal';
import '@testing-library/jest-dom';

jest.mock('crypto', () => ({ randomUUID: () => '123456789' }));
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(() => ({ push: jest.fn() })),
}));

describe('AddCarePartnerModal', () => {
  it('should render', () => {
    const { container } = render(
      <AddCarePartnerModal
        isModalOpen
        setIsModalOpen={() => {}}
        onSubmit={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should prompt user when attempting to close modal with unsaved changes', () => {
    render(
      <AddCarePartnerModal
        isModalOpen
        setIsModalOpen={() => {}}
        onSubmit={() => {}}
      />
    );

    const nameInput = screen.getByLabelText(copyTexts.AddCarePartnerModal.name);

    fireEvent.change(nameInput, { target: { value: '123' } });

    expect(screen.queryByTestId('care-partner-close-warning')).toBeNull();

    const cancelButton = screen.getByText(
      copyTexts.AddCarePartnerModal.cancelButton
    );
    fireEvent.click(cancelButton);

    expect(screen.queryByTestId('care-partner-close-warning')).not.toBeNull();
  });

  it('should allow user to close after prompting about unsaved changes', () => {
    const setIsOpenOpenMock = jest.fn();
    render(
      <AddCarePartnerModal
        isModalOpen
        setIsModalOpen={setIsOpenOpenMock}
        onSubmit={() => {}}
      />
    );

    const nameInput = screen.getByLabelText(copyTexts.AddCarePartnerModal.name);

    fireEvent.change(nameInput, { target: { value: 'test' } });

    const cancelButton = screen.getByText(
      copyTexts.AddCarePartnerModal.cancelButton
    );
    fireEvent.click(cancelButton);

    const closeAnywaybutton = screen.getByText(
      copyTexts.AddCarePartnerModal.closeAnywayButton
    );
    fireEvent.click(closeAnywaybutton);

    expect(setIsOpenOpenMock).toHaveBeenCalledWith(false);
  });

  it('should call onSubmit when user clicks save', async () => {
    const onSubmitMock = jest.fn();
    render(
      <AddCarePartnerModal
        isModalOpen
        setIsModalOpen={() => {}}
        onSubmit={onSubmitMock}
      />
    );

    const nameInput = screen.getByLabelText(copyTexts.AddCarePartnerModal.name);
    fireEvent.change(nameInput, { target: { value: 'Mark' } });

    const emailInput = screen.getByLabelText(
      copyTexts.AddCarePartnerModal.email
    );
    fireEvent.change(emailInput, { target: { value: 'mark.smith@gmail.com' } });

    const saveButton = screen.getByText(
      copyTexts.AddCarePartnerModal.saveButton
    );
    fireEvent.click(saveButton);

    await waitFor(() => {
      expect(onSubmitMock).toHaveBeenCalledWith({
        emailAddress: 'mark.smith@gmail.com',
        name: 'Mark',
      });
    });
  });

  it('should render errors when save clicked with empty inputs', async () => {
    const { container } = render(
      <AddCarePartnerModal
        isModalOpen
        setIsModalOpen={() => {}}
        onSubmit={() => {}}
      />
    );

    const emailInput = screen.getByLabelText(
      copyTexts.AddCarePartnerModal.email
    );
    fireEvent.change(emailInput, { target: { value: 'mark' } });

    const saveButton = screen.getByText(
      copyTexts.AddCarePartnerModal.saveButton
    );
    fireEvent.click(saveButton);

    await waitFor(() => {
      expect(
        screen.queryByText(
          copyTexts.AddCarePartnerModal.formValidation.invalidEmail
        )
      ).toBeInTheDocument();
      expect(
        screen.queryByText(
          copyTexts.AddCarePartnerModal.formValidation.requiredField
        )
      ).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });
});
