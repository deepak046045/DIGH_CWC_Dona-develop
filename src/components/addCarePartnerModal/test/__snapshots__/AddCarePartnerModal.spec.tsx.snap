// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCarePartnerModal should render 1`] = `
<div>
  <div
    class="primary modal"
  >
    <div
      class="lds-modal open"
      id="addCarePartnerModal"
    >
      <div
        class="wrapper"
      >
        <div
          class="row justify-content-center"
        >
          <div
            class="col-12 col-md-9"
          >
            <div
              aria-labelledby=""
              aria-modal="true"
              class="lds-modal-window"
              role="dialog"
            >
              <div
                class="lds-modal-body"
              >
                <div
                  class="closeModalButtonWrapper"
                >
                  <button
                    class="closeModalButton"
                    type="button"
                  >
                    <svg
                      aria-labelledby="lds-icon-X-title-123456789 lds-icon-X-description-123456789"
                      class="lds-icon icon-X closeIcon"
                      focusable="false"
                      role="img"
                    >
                      <title
                        id="lds-icon-X-title-123456789"
                      >
                        X
                      </title>
                      <desc
                        id="lds-icon-X-description-123456789"
                      />
                      <svg
                        viewBox="0 0 256 256"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M205.66,194.34a8,8,0,0,1-11.32,11.32L128,139.31,61.66,205.66a8,8,0,0,1-11.32-11.32L116.69,128,50.34,61.66A8,8,0,0,1,61.66,50.34L128,116.69l66.34-66.35a8,8,0,0,1,11.32,11.32L139.31,128Z"
                        />
                      </svg>
                    </svg>
                  </button>
                </div>
                <h2
                  class="modalTitle"
                >
                  Add care partner
                </h2>
                <form
                  class="form"
                >
                  <div
                    class="formContent"
                  >
                    <div>
                      <label
                        class="lds-form-label"
                      >
                        NAME
                        <input
                          aria-describedby=""
                          class="lds-text-field"
                          id="name"
                          name="name"
                          type="text"
                          value=""
                        />
                      </label>
                    </div>
                    <div>
                      <label
                        class="lds-form-label"
                      >
                        EMAIL ADDRESS
                        <input
                          aria-describedby=""
                          class="lds-text-field"
                          id="emailAddress"
                          name="emailAddress"
                          type="text"
                          value=""
                        />
                      </label>
                    </div>
                  </div>
                  <div
                    class="modalActions"
                  >
                    <button
                      class="lds-button secondary"
                      type="button"
                    >
                      Cancel
                    </button>
                    <button
                      class="lds-button primary"
                      data-testid="add-care-partner-save-button"
                      type="submit"
                    >
                      Save
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AddCarePartnerModal should render errors when save clicked with empty inputs 1`] = `
<div>
  <div
    class="primary modal"
  >
    <div
      class="lds-modal open"
      id="addCarePartnerModal"
    >
      <div
        class="wrapper"
      >
        <div
          class="row justify-content-center"
        >
          <div
            class="col-12 col-md-9"
          >
            <div
              aria-labelledby=""
              aria-modal="true"
              class="lds-modal-window"
              role="dialog"
            >
              <div
                class="lds-modal-body"
              >
                <div
                  class="closeModalButtonWrapper"
                >
                  <button
                    class="closeModalButton"
                    type="button"
                  >
                    <svg
                      aria-labelledby="lds-icon-X-title-123456789 lds-icon-X-description-123456789"
                      class="lds-icon icon-X closeIcon"
                      focusable="false"
                      role="img"
                    >
                      <title
                        id="lds-icon-X-title-123456789"
                      >
                        X
                      </title>
                      <desc
                        id="lds-icon-X-description-123456789"
                      />
                      <svg
                        viewBox="0 0 256 256"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M205.66,194.34a8,8,0,0,1-11.32,11.32L128,139.31,61.66,205.66a8,8,0,0,1-11.32-11.32L116.69,128,50.34,61.66A8,8,0,0,1,61.66,50.34L128,116.69l66.34-66.35a8,8,0,0,1,11.32,11.32L139.31,128Z"
                        />
                      </svg>
                    </svg>
                  </button>
                </div>
                <h2
                  class="modalTitle"
                >
                  Add care partner
                </h2>
                <form
                  class="form"
                >
                  <div
                    class="formContent"
                  >
                    <div>
                      <label
                        class="lds-form-label"
                      >
                        NAME
                        <input
                          aria-describedby=""
                          aria-invalid="true"
                          class="lds-text-field error"
                          id="name"
                          name="name"
                          type="text"
                          value=""
                        />
                      </label>
                      <span
                        class="lds-form-validation-error"
                        id="errName"
                      >
                        <lds-icon
                          decorative="true"
                          inline="true"
                          name="WarningCircleFill"
                        />
                         
                        This field is required
                      </span>
                    </div>
                    <div>
                      <label
                        class="lds-form-label"
                      >
                        EMAIL ADDRESS
                        <input
                          aria-describedby=""
                          aria-invalid="true"
                          class="lds-text-field error"
                          id="emailAddress"
                          name="emailAddress"
                          type="text"
                          value="mark"
                        />
                      </label>
                      <span
                        class="lds-form-validation-error"
                        id="errEmailAddress"
                      >
                        <lds-icon
                          decorative="true"
                          inline="true"
                          name="WarningCircleFill"
                        />
                         
                        Invalid email address
                      </span>
                    </div>
                  </div>
                  <div
                    class="modalActions"
                  >
                    <button
                      class="lds-button secondary"
                      type="button"
                    >
                      Cancel
                    </button>
                    <button
                      class="lds-button primary"
                      data-testid="add-care-partner-save-button"
                      type="submit"
                    >
                      Save
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
